x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termenum/action/TransportTermEnumAction.java

```diff
@@ -0,0 +1,552 @@
+                if (System.currentTimeMillis() > scheduledEnd) {
+                    return new NodeTermEnumResponse(request.nodeId(), termsList, error, false);
+                }
+                final IndexService indexService = indicesService.indexServiceSafe(shardId.getIndex());
+                final IndexShard indexShard = indexService.getShard(shardId.getId());
+
+                Engine.Searcher searcher = indexShard.acquireSearcher(Engine.SEARCH_SOURCE);
+                openedResources.add(searcher);
+                final SearchExecutionContext queryShardContext = indexService.newSearchExecutionContext(
+                    shardId.id(),
+                    0,
+                    searcher,
+                    request::shardStartedTimeMillis,
+                    null,
+                    Collections.emptyMap()
+                );
+                final MappedFieldType mappedFieldType = indexShard.mapperService().fieldType(request.field());
+                if (mappedFieldType != null) {
+                    TermsEnum terms = mappedFieldType.getTerms(request.caseInsensitive(), request.string(), queryShardContext);
+                    if (terms != null) {
+                        shardTermsEnums.add(terms);
+                    }
+                }
+            }
+            MultiShardTermsEnum te = new MultiShardTermsEnum(shardTermsEnums.toArray(new TermsEnum[0]));
+
+            int shard_size = request.size();
+            // All the above prep might take a while - do a timer check now before we continue further.
+            if (System.currentTimeMillis() > scheduledEnd) {
+                return new NodeTermEnumResponse(request.nodeId(), termsList, error, false);
+            }
+
+            int numTermsBetweenClockChecks = 100;
+            int termCount = 0;
+            // Collect in alphabetical order
+            while (te.next() != null) {
+                termCount++;
+                if (termCount > numTermsBetweenClockChecks) {
+                    if (System.currentTimeMillis() > scheduledEnd) {
+                        boolean complete = te.next() == null;
+                        return new NodeTermEnumResponse(request.nodeId(), termsList, error, complete);
+                    }
+                    termCount = 0;
+                }
+                long df = te.docFreq();
+                BytesRef bytes = te.term();
+                termsList.add(new TermCount(bytes.utf8ToString(), df));
+                if (termsList.size() >= shard_size) {
+                    break;
+                }
+            }
+
+        } catch (Exception e) {
+            error = e.getMessage();
+        } finally {
+            IOUtils.close(openedResources);
+        }
+        return new NodeTermEnumResponse(request.nodeId(), termsList, error, true);
+    }
+
+    // TODO remove this so we can shift code to server module - see https://github.com/elastic/elasticsearch/issues/70221
```

Comment:
Is this still accurate. I thought that we would try to rewrite the role query and if it results in a match_all then we'd accept the request ?

Reply:
The TODO should perhaps be about moving any security filtering logic like this to a custom "Interceptor" class which wraps the core implementation and rewrites requests rather than having security+impl in the same class.